### PR TITLE
List support, CKM_RSA_X_509

### PR DESCRIPTION
**Pull request by Emantor**
This PR fixes list support using pkcs11-tool, since without `req_attrs` set the objects would not be opened.
There is also support for `CKM_RSA_X_509` to support sign and verify for RSA without padding. This allows the use of the openssl pkcs11 engine module with SKS and client or server certificates.

### CODE REVIEW

File: ta/secure_key_services/src/object.c [hunk 1]
```diff
@@ -393,6 +375,34 @@ static uint32_t token_obj_matches_ref(struct sks_attrs_head *req_attrs,
...
+
+	rv = open_object(obj);
+	if(rv)
+		return rv;
+
+	if (!attributes_match_reference(obj->attributes, req_attrs)) {
+		return SKS_NOT_FOUND;
+	}
+
+	rv = SKS_OK;
```
Comment (etienne-lms) [inline ta/secure_key_services/src/object.c]: minor: `return rv;` only is fine

File: lib/libutee/tee_api_operations.c [hunk 2]
```diff
@@ -259,6 +259,10 @@ TEE_Result TEE_AllocateOperation(TEE_OperationHandle *operation,
 #ifdef CFG_CRYPTO_RSASSA_NA1
 	if (algorithm == TEE_ALG_RSASSA_PKCS1_V1_5)
 		op->info.operationClass = TEE_OPERATION_ASYMMETRIC_SIGNATURE;
+#endif
+#ifdef CFG_CRYPTO_RSA_NOPAD_SIGNATURE
+	if (mode == TEE_MODE_SIGN || mode == TEE_MODE_VERIFY)
```
Comment (etienne-lms) [inline lib/libutee/tee_api_operations.c]:
```
if (algorithm == TEE_ALG_RSA_NOPAD &&
   (mode == TEE_MODE_SIGN || mode == TEE_MODE_VERIFY))
       ...
```

**Review by etienne-lms — COMMENTED** [review]
Hi @Emantor, please find few comments. The whole looks sane to me and the generic commit could be submitted to OP-TEE OS.

File: lib/libutee/tee_api_operations.c (hunk 2 above)
Comment (Emantor) [reply lib/libutee/tee_api_operations.c]: Oh, of course, I think I lost this during refactoring. Thanks :+1:

**Comment by Emantor** [comment]
Updated.

File: ta/secure_key_services/src/object.c [hunk 3]
```diff
@@ -393,6 +375,32 @@ static uint32_t token_obj_matches_ref(struct sks_attrs_head *req_attrs,
...
+	uint32_t rv = SKS_OK;
+	TEE_ObjectInfo info;
+
+	TEE_MemFill(&info, 0, sizeof(info));
+
+	if (obj->attributes) {
+		if (!attributes_match_reference(obj->attributes, req_attrs))
+			return SKS_NOT_FOUND;
+
+		return rv;
```
Comment (etienne-lms) [inline ta/secure_key_services/src/object.c:390]: minor: maybe simplify (and remove unused `info`):
```
	if (!obj->attributes) {
		if (open_object(obj))
			return SKS_GENERIC;
	}

	if (!attributes_match_reference(obj->attributes, req_attrs))
		return SKS_NOT_FOUND;

	return SKS_OK;
}
```

File: ta/secure_key_services/src/object.c [hunk 4]
```diff
@@ -393,6 +375,32 @@ static uint32_t token_obj_matches_ref(struct sks_attrs_head *req_attrs,
 	return rv;
 }
 
+static uint32_t token_obj_matches_ref(struct sks_attrs_head *req_attrs,
+				      struct sks_object *obj)
+{
+	uint32_t rv = SKS_OK;
+	TEE_ObjectInfo info;
+
+	TEE_MemFill(&info, 0, sizeof(info));
```
Comment (etienne-lms) [inline ta/secure_key_services/src/object.c:384]: remove unused `info`

File: ta/secure_key_services/src/object.c [hunk 5]
```diff
@@ -310,25 +310,14 @@ uint32_t entry_destroy_object(uintptr_t tee_session, TEE_Param *ctrl,
...
 }
 
-static uint32_t token_obj_matches_ref(struct sks_attrs_head *req_attrs,
-				      struct sks_object *obj)
-{
+static uint32_t open_object(struct sks_object *obj) {
 	uint32_t rv = 0;
 	TEE_Result res = TEE_ERROR_GENERIC;
 	TEE_ObjectHandle hdl = obj->attribs_hdl;
 	TEE_ObjectInfo info;
```
Comment (etienne-lms) [inline ta/secure_key_services/src/object.c:317]: `TEE_ObjectInfo info = { };` (cleaner, and required since change below)

**Review by etienne-lms — COMMENTED** [review]
Late comments for commit "sks ta: open object if no req_attrs".

**Pull request closed without merging**